### PR TITLE
Ghc8 support, take 2.

### LINKED ISSUE #2: GHC 8 support
The GHC backend now supports GHC 7.10 AST.

The code should be adapted so it keeps supporting that, as well the GHC 8 AST.

Note: while there, backwards support for the GHC 7.8 AST would be very easy, since there was only a slight change.

### PR DESCRIPTION
**Pull request by robinp**
Tried to keep changes minimum for other pending work to merge more smoothly. Once things settled, adding a few helpers could get help rid of most of the ifdefs.

Fixes #2. +@gleber for review, cc +kristoff3r, +ivan444.

### CODE REVIEW

**Comment by robinp** [comment]
Note: @kristoff3r already agreed on contributing to this project when he committed to the ghc8 branch. This PR just merges that to the master branch. Christopher, feel free to object :)

File: haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs [hunk 1]
```diff
@@ -507,58 +620,98 @@ deepDeclsFromTopBind ctx top =
...
-    let (abss, funs) = partitionEithers
-                     . concatMap (absFunDeclsFromHsBind ctx)
-                     $ binds
-        -- FunBind will have the same funVar as the monomorphic binding of the
-        -- AbsBind (if there's a corresponding AbsBind).
-        absMonoTicks = mapMaybe daAlt abss
-        funOnlys = filter (not . hasTick absMonoTicks) funs
-    in map (flip DeclAndAlt Nothing) funOnlys ++ abss
+declsFromHsBinds :: ExtractCtx -> [ParentChild (LHsBind Id)] -> [DeclAndAlt]
+declsFromHsBinds ctx = concatMap thing
```
Comment (gleber) [inline haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs]: s/thing/go/ ?

**Comment by kristoff3r** [comment]
I agree to my commits being used for this project.

**Comment by kristoff3r** [comment]
Also thanks for finishing my pull request, I'm glad I didn't have to do the CPP spaghetti :D

File: haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs [hunk 2]
```diff
@@ -827,20 +1007,41 @@ refsFromTypechecked ctx tsrc declAlts =
         -- The same-named fields of different datacons are actually just a
         -- single field, so it's ok to pass in an arbitrary DataCon to mine
         -- for reference targets.
+        --recordUpdRefs :: DataCon -> [LHsRecUpdField Id]
```
Comment (gleber) [inline haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs]: Either remove or add signature based on GHC version

**Pull request merged**